openapi3filter/issue884_test.go

```diff
@@ -0,0 +1,113 @@
+package openapi3filter
+
+import (
+	"net/http"
+	"net/url"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/getkin/kin-openapi/openapi3"
+	"github.com/getkin/kin-openapi/routers/gorillamux"
+)
+
+func TestIssue884(t *testing.T) {
+	loader := openapi3.NewLoader()
+	ctx := loader.Context
+	spec := `
+  openapi: 3.0.0
+  info:
+    version: 1.0.0
+    title: Sample API
+  components:
+    schemas:
+      TaskSortEnum:
+        enum:
+          - createdAt
+          - -createdAt
+          - updatedAt
+          - -updatedAt
+  paths:
+    /tasks:
+      get:
+        operationId: ListTask
+        parameters:
+          - in: query
+            name: withDefault
+            schema:
+              allOf:
+                - $ref: '#/components/schemas/TaskSortEnum'
+                - default: -createdAt
+          - in: query
+            name: withoutDefault
+            schema:
+              allOf:
+                - $ref: '#/components/schemas/TaskSortEnum'
+          - in: query
+            name: withManyDefaults
+            schema:
+              allOf:
+                - default: -updatedAt
+                - $ref: '#/components/schemas/TaskSortEnum'
+                - default: -createdAt
+        responses:
+          '200':
+            description: Successful response
+  `[1:]
+
+	doc, err := loader.LoadFromData([]byte(spec))
+	require.NoError(t, err)
+
+	err = doc.Validate(ctx)
+	require.NoError(t, err)
+
+	router, err := gorillamux.NewRouter(doc)
+	require.NoError(t, err)
+
+	tests := []struct {
+		name          string
+		options       *Options
+		expectedQuery url.Values
+	}{
+		{
+			name: "no defaults are added to requests",
+			options: &Options{
+				SkipSettingDefaults: true,
+			},
+			expectedQuery: url.Values{},
+		},
+
+		{
+			name: "defaults are added to requests",
+			expectedQuery: url.Values{
+				"withDefault":      []string{"-createdAt"},
+				"withManyDefaults": []string{"-updatedAt"}, // first default is win
```

Comment:
Come to think of it, it would be more natural for the last value to be win when schema validation errors is not raised?

Reply:
To be honest, I don't know which choice is better.

- The first value is win (low cost)
- The last value is win (more natural)

Reply:
( In oapi-codegen, the allOf has multiple default values ​​will cause an error when merging schema phase, but this is oapi-codegen's story. )

Reply:
AFAICT the spec is blurry here. 
So to me either choice is fine as long as it's mentioned in the field docs. 

My vote for using the earliest default for now (shorter code path). When the need arrives, then we can pick another way to read this, like a func for instance. 
You decide :)

Reply:
OK I also like "first default is win".

(As an extension of the "last default is win" specification, if deep nested structures were to be supported, a complete full scan would be required)